### PR TITLE
Fix to allow unicode py2.7

### PR DESCRIPTION
**Pull request by jklymak**
This is to fix https://github.com/nucleic/kiwi/issues/39.  It works on the test I have.  I admit I didn't run the other tests.  

It basically allows `from __future__ import unicode_literals` to be used in python 2.7.  I tested in Python 2.7 with and with out that line at the top for the test, and in python 3 w/ that line.  

```python
from __future__ import unicode_literals

import kiwisolver as kiwi

Variable = kiwi.Variable
solver = kiwi.Solver()
top = Variable('boo')
c = (top == 1.0)
solver.addConstraint(c | 'strong')
solver.dump()
```

@sccolbert @MatthieuDartiailh 

Thanks!

### CODE REVIEW

File: py/pythonhelpers.h [hunk 1]
```diff
@@ -11,15 +11,15 @@
...
-#define FROM_STRING PyUnicode_FromString
 #define INITERROR return NULL
 #define MOD_INIT_FUNC(name) PyMODINIT_FUNC PyInit_##name(void)
 #else
-#define FROM_STRING PyString_FromString
 #define INITERROR return
 #define MOD_INIT_FUNC(name) PyMODINIT_FUNC init##name(void)
 #endif
 
+#define FROM_STRING PyUnicode_FromString
```
Comment (sccolbert) [inline py/pythonhelpers.h]: Why did you make this change? Returning `unicode` in Py27 where it was previously `str` is an API-incompatible change.

File: py/util.h [hunk 2]
```diff
@@ -44,14 +44,19 @@ convert_to_double( PyObject* obj, double& out )
...
+  std::string str;
+
 #if PY_MAJOR_VERSION >= 3
     if( PyUnicode_Check( value ) )
     {
-        std::string str( PyUnicode_AsUTF8( value ) );
+      str = PyUnicode_AsUTF8( value );
 #else
-    if( PyString_Check( value ) )
+    if( PyString_Check( value ) | PyUnicode_Check( value ))
```
Comment (sccolbert) [inline py/util.h]: I would make this and `if { } else if {} else {}` block instead of nested `if`

Comment (jklymak) [reply py/util.h]: I agree it looks a bit inelegant, and I'm not the most elegant programmer, so no offense if you or someone else changes it.   But, you need an outer `if` to check for a string to send to the string translators below, and then you need to decide which string decoder you want.

Comment (MatthieuDartiailh) [reply py/util.h]: I am not sure it would make this any more elegant as we would have either to duplicate the following logic (string value comparison) or refactor it in a function.  To me this looks fine.

File: py/util.h [hunk 3]
```diff
@@ -44,14 +44,19 @@ convert_to_double( PyObject* obj, double& out )
...
     {
-        std::string str( PyString_AS_STRING( value ) );
+      if( PyUnicode_Check( value ) )
+          str = PyString_AS_STRING(PyUnicode_AsASCIIString( value ) );
```
Comment (MatthieuDartiailh) [inline py/util.h]: Why not use PyUnicode_AsUTF8 as we do on Python 3 ?

**Comment by MatthieuDartiailh** [comment]
To me this looks mostly good to go. I like in particular the fact it actually remove some lines of C : )

File: py/util.h (hunk 2 above)
Comment (sccolbert) [reply py/util.h]: Yep. You're both right. I only looked at the diff preview and was missing the extra context.

File: py/util.h (hunk 3 above)
Comment (sccolbert) [reply py/util.h]: Yeah. `PyUnicode_AsASCIIString` also returns a new object, so this line will cause a memory leak.
https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsASCIIString

**Comment by MatthieuDartiailh** [comment]
@sccolbert Are you happy with it ? After merging this, we could tag a 1.0.1 as if this goes into matplotlib they will want a tag.

**Comment by sccolbert** [comment]
@MatthieuDartiailh Not quite happy with it yet, because of the memory leak I mentioned above.

**Comment by jklymak** [comment]
@MatthieuDartiailh   No rush on the tag if you and @sccolbert want to accumulate other feedback.  It fails matplotlib for python 2.7, but does fine for 3.6, which I think most of the devs are on.  The changes I'm making in matplotlib will take a good while to get out of development, and I know it works in 2.7 so failing that test for a while (I'd bet a couple of months at least) should be fine.

**Comment by jklymak** [comment]
BTW, thanks for fixing my python/C/Unicode ignorance.   Its been a long time since I coded in C, I'm mostly just a hacker in python, and I have only a vague understanding that encoding text is a pain.

FWIW, the `matplotlib`work I'm doing just needs ASCII (users aren't exposed to the strings), but I can imagine other users might try unicode.

**Comment by jklymak** [comment]
BTW this works fine w/ my test.  Not sure if you want to add that test.  There are probably a good chunk of py2/3 projects that use the unicode future...

**Comment by MatthieuDartiailh** [comment]
Now that I am done messing up the C code .... I will add a test by simply forcing a string to be unicode using the u prefix.

**Comment by MatthieuDartiailh** [comment]
@jklymak you forgot the setName method of variable. I fixed that and updated the tests. @sccolbert is this now good to go for you ?

**Comment by MatthieuDartiailh** [comment]
Fixed the possible segfault under Python 2

**Comment by MatthieuDartiailh** [comment]
This is ready for review.

**Comment by jklymak** [comment]
@MatthieuDartiailh  Would you like me to rebase this so it is just one commit?  `matplotlib` often asks for this just to keep the commit tree clean.  I think it makes `blame` a lot easier if its one commit per pull request merge).  OTOH if you want the history above to make perfect sense, then rebasing is a less good idea....

**Comment by MatthieuDartiailh** [comment]
As you wish. Nowadays I use the Squash & Merge option of github so that people don't need to squash manually.

**Comment by jklymak** [comment]
@MatthieuDartiailh   Not having managed a codebase before I didn't know that existed.  Very useful!  I'll let you guys deal with it on merge.

File: py/util.h [hunk 4]
```diff
@@ -44,14 +44,27 @@ convert_to_double( PyObject* obj, double& out )
...
+    PyObject* ascii_str;
+
+    if( PyString_Check( value ) | PyUnicode_Check( value ))
     {
-        std::string str( PyString_AS_STRING( value ) );
+      if( PyUnicode_Check( value ) )
+      {
+          ascii_str = PyUnicode_AsASCIIString( value );
+          if( !ascii_str )
+              return 0;
```
[the remark this answers is not in the document]
Comment (sccolbert) [reply py/util.h]: This block should probably use the UTF8 encoding just like we do for py3. You can make use of a smart pointer so you don't need to manage the ref count manually. I would remove the `PyObject* ascii_str;` declaration, and use a block like this:

```
          PythonHelpers::PyObjectPtr py_str( PyUnicode_AsUTF8String( value ) );
          if( !py_str )
              return false;
          str = PyString_AS_STRING( py_str.get() );
```

File: py/util.h [hunk 5]
```diff
@@ -44,14 +44,27 @@ convert_to_double( PyObject* obj, double& out )
 inline bool
 convert_to_strength( PyObject* value, double& out )
 {
+    std::string str;
```
Comment (sccolbert) [inline py/util.h]: This variable declaration does not need to be hoisted.

File: py/util.h [hunk 6]
```diff
@@ -44,14 +44,27 @@ convert_to_double( PyObject* obj, double& out )
...
+
 #if PY_MAJOR_VERSION >= 3
     if( PyUnicode_Check( value ) )
     {
-        std::string str( PyUnicode_AsUTF8( value ) );
+      str = PyUnicode_AsUTF8( value );
```
Comment (sccolbert) [inline py/util.h]: this line did not need to be changed.

File: py/util.h [hunk 7]
```diff
@@ -44,14 +44,27 @@ convert_to_double( PyObject* obj, double& out )
...
 #else
-    if( PyString_Check( value ) )
+    PyObject* ascii_str;
+
+    if( PyString_Check( value ) | PyUnicode_Check( value ))
     {
-        std::string str( PyString_AS_STRING( value ) );
```
Comment (sccolbert) [inline py/util.h]: This should be changed to `std::string str;`

File: py/util.h [hunk 8]
```diff
@@ -44,14 +44,27 @@ convert_to_double( PyObject* obj, double& out )
...
+    std::string str;
+
 #if PY_MAJOR_VERSION >= 3
     if( PyUnicode_Check( value ) )
     {
-        std::string str( PyUnicode_AsUTF8( value ) );
+      str = PyUnicode_AsUTF8( value );
 #else
-    if( PyString_Check( value ) )
+    PyObject* ascii_str;
```
Comment (sccolbert) [inline py/util.h]: This is not needed. See longer comment below.

File: py/variable.cpp [hunk 9]
```diff
@@ -105,9 +94,25 @@ Variable_setName( Variable* self, PyObject* pystr )
...
+
+   if( PyString_Check( pystr ) | PyUnicode_Check( pystr ))
+   {
+     if( PyUnicode_Check( pystr ) )
+     {
+         ascii_str = PyUnicode_AsASCIIString( pystr );
+         if( !ascii_str )
+             return 0;
+         str = PyString_AS_STRING( ascii_str );
+         Py_DECREF( ascii_str );
```
Comment (sccolbert) [inline py/variable.cpp]: This block can be done the same way as above with a smart pointer.

File: py/tests/test_constraint.py [hunk 10]
```diff
@@ -27,6 +30,10 @@ def test_constraint_creation():
         c = Constraint(v + 1, '>=', s)
         assert c.strength() == getattr(strength, s)
 
+    if sys.version_info < (3,):
+        with pytest.raises(UnicodeEncodeError):
```
Comment (ChrisBarker-NOAA) [inline py/tests/test_constraint.py]: in py2, folks can very often accidentally use a mixture of Unicode and plain strings. It would be best if that was allowed.

Ideally, you convert all py2 strings into unicode at the calling boundary, and then expect everything to be unicode internally.

I haven't dug into your C++ code (why in the world are you hand-writing the bindings??? -- Cython would make this a lot easier) -- but a little utility conversion function should be fairly easy to write and drop in everywhere a string is expected.

It looks like this is testing for non-ascii character -- but if it's allowed in py3, why not py2??

Comment (sccolbert) [reply py/tests/test_constraint.py]: > I haven't dug into your C++ code (why in the world are you hand-writing the bindings??? -- Cython would make this a lot easier) 

It really doesn't, and this sort of comment is not helpful.

Comment (ChrisBarker-NOAA) [reply py/tests/test_constraint.py]: no, it's not -- sorry.

But in any case, I have now looked at the C++ and it looks like you want a utf-8 encoded std::string from a python string.

In python, you would do this by calling .encode('utf-8') on the object, and that would work on py2 strings and unicode objects, and py3 strings. So maybe you could call the python from C++, but that's awkward, and maybe slow, so you could instead call:

PyString_Encode on py2 strings, and 

PyUnicode_AsUTF8String on py2 unicode objects and py3 strings.

In any case, why require ascii on py2 and allow full unicode on py3?

Comment (MatthieuDartiailh) [reply py/tests/test_constraint.py]: Sorry my mistake. I chose to stay close to the original contributor implementation. I will fix this.

**Comment by MatthieuDartiailh** [comment]
After much more tinkering with the C-API I managed to get Python 2 to accept any unicode name for a variable (both at creation and when setting it) and accept unicode for strength and constraint operator.
I added a new helper function in util.h to handle the conversion and avoid code duplication.

**Comment by sccolbert** [comment]
This is a really good update, but there's one more place we need to be defensive with the new changes. I'm going to approve my old change request and make a new comment on the new diff.

File: py/variable.cpp [hunk 11]
```diff
@@ -44,10 +37,18 @@ Variable_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
...
-    	new( &self->variable ) kiwi::Variable( PyString_AS_STRING( name ) );
+      if( !(PyString_Check( name ) | PyUnicode_Check( name ) ) )
+      {
+         return py_expected_type_fail( name, "str or unicode" );
+      }
 #endif
+      std::string c_name;
+      if( !convert_pystr_to_str(name, c_name) )
+          return 0;
+    	new( &self->variable ) kiwi::Variable( c_name );
```
Comment (sccolbert) [inline py/variable.cpp]: Since this change causes the function to possible return early after the python variable object has been allocated, we need to use a smart pointer to guard the lifetime of that object. Here's an updated version of this function which does that:

```c++
static PyObject*
Variable_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
{
	static const char *kwlist[] = { "name", "context", 0 };
	PyObject* context = 0;
	PyObject* name = 0;

	if( !PyArg_ParseTupleAndKeywords(
		args, kwargs, "|OO:__new__", const_cast<char**>( kwlist ),
		&name, &context ) )
		return 0;

	PyObjectPtr pyvar( PyType_GenericNew( type, args, kwargs ) );
	if( !pyvar )
		return 0;

	Variable* self = reinterpret_cast<Variable*>( pyvar.get() );
	self->context = xnewref( context );

	if( name != 0 )
	{
#if PY_MAJOR_VERSION >= 3
		if( !PyUnicode_Check( name ) )
			return py_expected_type_fail( name, "unicode" );
#else
		if( !( PyString_Check( name ) | PyUnicode_Check( name ) ) )
		{
			return py_expected_type_fail( name, "str or unicode" );
		}
#endif
		std::string c_name;
		if( !convert_pystr_to_str(name, c_name) )
			return 0;
		new( &self->variable ) kiwi::Variable( c_name );
	}
	else
	{
		new( &self->variable ) kiwi::Variable();
	}

	return pyvar.release();
}
```

Comment (MatthieuDartiailh) [reply py/variable.cpp]: Thanks for catching this.

**Comment by sccolbert** [comment]
Thanks!

**Comment by MatthieuDartiailh** [comment]
A squash and merge would have been cleaner for the history ( I tinkered a lot and used Travis for Python 2.7 feedback) but it does not really matter.

**Comment by jklymak** [comment]
@MatthieuDartiailh @scolbert

Would it be possible to issue a release based on this change?  That way I can get #9082 in matplotlib merged.

**Comment by MatthieuDartiailh** [comment]
I will try to do one this week. Ping me if I forget.

**Comment by MatthieuDartiailh** [comment]
The release is done

**Comment by jklymak** [comment]
Thanks!

**Comment by MatthieuDartiailh** [comment]
In case you need it the conda-forge update is also on its way.

**Comment by jklymak** [comment]
Thanks so much!  Seems to be working great.
